pkg/networkservice/common/heal/client.go

```diff
@@ -67,43 +68,85 @@ func NewClient(client networkservice.MonitorConnectionClient, onHeal *networkser
 		requestors:        make(map[string]func()),
 		closers:           make(map[string]func()),
 		reported:          make(map[string]*networkservice.Connection),
+		connectionIDs:     make(map[string]bool),
 		client:            client,
 		updateExecutor:    serialize.NewExecutor(),
 		eventReceiver:     nil, // This is intentionally nil
 		recvEventExecutor: serialize.NewExecutor(),
 	}
+
 	if rv.onHeal == nil {
 		rv.onHeal = addressof.NetworkServiceClient(rv)
 	}
-	rv.updateExecutor.AsyncExec(func() {
-		runtime.SetFinalizer(rv, func(f *healClient) {
-			f.updateExecutor.AsyncExec(func() {
-				if f.cancelFunc != nil {
-					f.cancelFunc()
-				}
-			})
-		})
-		ctx, cancelFunc := context.WithCancel(context.Background())
-		rv.cancelFunc = cancelFunc
-		// TODO decide what to do about err here
-		recv, _ := rv.client.MonitorConnections(ctx, nil)
-		rv.eventReceiver = recv
-		rv.recvEventExecutor.AsyncExec(rv.recvEvent)
-	})
 
 	return rv
 }
 
+func (f *healClient) init() {
+	if f.eventReceiver != nil {
+		select {
+		case <-f.eventReceiver.Context().Done():
+			return
+		default:
+			f.eventReceiver = nil
+		}
+	}
+
+	logrus.Info("Creating new eventReceiver")
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	f.cancelFunc = cancelFunc
+
+	recv, err := f.client.MonitorConnections(ctx, &networkservice.MonitorScopeSelector{}, grpc.WaitForReady(true))
+	if err != nil {
+		f.updateExecutor.AsyncExec(f.init)
+		return
+	}
+
+	f.eventReceiver = recv
+	f.recvEventExecutor.AsyncExec(f.recvEvent)
+}
+
+func (f *healClient) lazyInit(id string) {
+	f.updateExecutor.SyncExec(func() {
+		if _, ok := f.connectionIDs[id]; !ok {
+			f.connectionIDs[id] = true
+			f.updateExecutor.AsyncExec(f.init)
+		}
+	})
+}
+
+func (f *healClient) tearDown() {
+	if f.cancelFunc != nil {
```

Comment:
Why are we calling cancelFunc in tearDown which is called by lazyTeardown which is called by Close ... this will cancelFunc the monitor for every closed Connection... don't we want to keep the MonitorClient going for the lifetime of the heal.NewClient struct... rather than just a particular connection?

Reply:
Fixed with eliminating of laziness